rotors_gazebo_plugins/src/gazebo_noisydepth_plugin.cpp

```diff
@@ -0,0 +1,308 @@
+                                       unsigned int _width, unsigned int _height, unsigned int _depth,
+                                       const std::string &_format) {
+  if (!this->initialized_ || this->height_ <= 0 || this->width_ <= 0)
+    return;
+
+  this->depth_sensor_update_time_ = this->parentSensor->LastMeasurementTime();
+  if (this->parentSensor->IsActive()) {
+    if (this->depth_image_connect_count_ <= 0 &&
+        (*this->image_connect_count_) <= 0) {
+      this->parentSensor->SetActive(false);
+    } else {
+      if (this->depth_image_connect_count_ > 0)
+        this->FillDepthImage(_image);
+    }
+  } else {
+    if (this->depth_image_connect_count_ <= 0)
+      // do this first so there's chance for sensor to run 1 frame after activate
+      this->parentSensor->SetActive(true);
+  }
+  PublishCameraInfo();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Update the controller
+void GazeboNoisyDepth::OnNewImageFrame(const unsigned char *_image,
+                                       unsigned int _width, unsigned int _height, unsigned int _depth,
+                                       const std::string &_format) {
+  if (!this->initialized_ || this->height_ <= 0 || this->width_ <= 0)
+    return;
+
+  //ROS_ERROR_NAMED("openni_kinect", "camera_ new frame %s %s",this->parentSensor_->Name().c_str(),this->frame_name_.c_str());
+  this->sensor_update_time_ = this->parentSensor_->LastMeasurementTime();
+
+  if (this->parentSensor->IsActive()) {
+    if (this->depth_image_connect_count_ <= 0 &&
+        (*this->image_connect_count_) <= 0) {
+      this->parentSensor->SetActive(false);
+    } else {
+      if ((*this->image_connect_count_) > 0)
+        this->PutCameraData(_image);
+    }
+  } else {
+    if ((*this->image_connect_count_) > 0)
+      // do this first so there's chance for sensor to run 1 frame after activate
+      this->parentSensor->SetActive(true);
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Put depth image data to the interface
+void GazeboNoisyDepth::FillDepthImage(const float *_src) {
+  this->lock_.lock();
+  // copy data into image
+  this->depth_image_msg_.header.frame_id = this->frame_name_;
+  this->depth_image_msg_.header.stamp.sec = this->depth_sensor_update_time_.sec;
+  this->depth_image_msg_.header.stamp.nsec = this->depth_sensor_update_time_.nsec;
+
+  ///copy from depth to depth image message
+  FillDepthImageHelper(this->depth_image_msg_,
+                       this->height,
+                       this->width,
+                       this->skip_,
+                       (void *) _src);
+
+  this->depth_image_pub_.publish(this->depth_image_msg_);
+
+  this->lock_.unlock();
+}
+
+// Fill depth information
+bool GazeboNoisyDepth::FillDepthImageHelper(
+    sensor_msgs::Image &image_msg,
+    uint32_t rows_arg, uint32_t cols_arg,
+    uint32_t step_arg, void *data_arg) {
```

Comment:
try to pass data_arg in by const reference and a proper type and move image_msgs to the end and switch to pointer type